### PR TITLE
Implement `new_files`

### LINKED ISSUE #1786: Implement `new_files`
https://github.com/microsoft/onefuzz/blob/febaf6e057aa8e14d8fdea5edb9590175458e00b/src/api-service/__app__/onefuzzlib/notifications/main.py#L126

We're stubbing this out for now because of the notifications templating issue

### PR DESCRIPTION
**Pull request by tevoinea**
Closes #1786 

List of files with reviewable content (aka: not just using change):

* src/ApiService/ApiService/ApiService.csproj
* src/ApiService/ApiService/OneFuzzTypes/Enums.cs
* src/ApiService/ApiService/OneFuzzTypes/Events.cs
* src/ApiService/ApiService/OneFuzzTypes/Model.cs
* src/ApiService/ApiService/Program.cs
* src/ApiService/ApiService/QueueFileChanges.cs
* src/ApiService/ApiService/onefuzzlib/Containers.cs
* src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs
* src/ApiService/ApiService/onefuzzlib/Reports.cs
* src/ApiService/ApiService/onefuzzlib/Storage.cs
* src/ApiService/ApiService/onefuzzlib/TaskOperations.cs
* src/ApiService/ApiService/packages.lock.json

### CODE REVIEW

File: src/ApiService/ApiService/onefuzzlib/Containers.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    private LogTracer _loggerTracer;
+    private IStorage _storage;
+    private ICreds _creds;
+    private ArmClient _armClient;
+    public Containers(ILogTracerFactory loggerFactory, IStorage storage, ICreds creds)
+    {
+        _loggerTracer = loggerFactory.MakeLogTracer(Guid.NewGuid());
+        _storage = storage;
+        _creds = creds;
+        _armClient = new ArmClient(credential: _creds.GetIdentity(), defaultSubscriptionId: _creds.GetSubcription());
```
Comment (stishkin) [inline src/ApiService/ApiService/onefuzzlib/Containers.cs:29]: I think we might be ok by having ArmClient to be singleton
It's supposed to renew auth behind the scenes automatically

**Review by stishkin — APPROVED** [review]

File: src/ApiService/ApiService/onefuzzlib/Containers.cs (hunk 1 above)
Comment (stishkin) [reply src/ApiService/ApiService/onefuzzlib/Containers.cs:29]: we can do this in some other PR

File: src/ApiService/ApiService/onefuzzlib/Reports.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            if (expectReports)
+            {
+                _log.Error($"unable to parse report ({filePath}): unicode decode of report failed - {e.Message} {e.StackTrace}");
+            }
+            return null;
+        }
+    }
+}
+
+public class RegressionReportOrReport
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/Reports.cs:94]: More than happy to replace this with https://github.com/mcintyre321/OneOf

File: src/ApiService/ApiService/onefuzzlib/Containers.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            return (await client.GetBlobClient(name).DownloadContentAsync())
+                .Value.Content.ToArray();
+        }
+        catch (RequestFailedException)
+        {
+            return null;
+        }
+    }
+
+    public BlobContainerClient? FindContainer(Container container, StorageType storageType)
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/Containers.cs:52]: NIT: This whole function could be rewritten cleaner in linq

File: src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            }
+        }
+
+        await foreach (var (task, containers) in GetQueueTasks())
+        {
+            if (containers.Contains(container.ContainerName))
+            {
+                _log.Info($"queuing input {container.ContainerName} {filename} {task.TaskId}");
+                var url = _containers.GetFileSasUrl(container, filename, StorageType.Corpus, read: true, delete: true);
+                await _queue.SendMessage(task.TaskId.ToString(), System.Text.Encoding.UTF8.GetBytes(url.ToString()), StorageType.Corpus);
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs:81]: It would be nice to have an override for _queue.SendMessage that took a string instead of bytes and did the conversion internally

File: src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        return _taskOperations.SearchStates(states: TaskStateHelper.Available())
+            .Select(task => (task, _taskOperations.GetInputContainerQueues(task.Config)))
+            .Where(taskTuple => taskTuple.Item2 != null)!;
+    }
+
+    private async Async.Task<Task?> GetRegressionReportTask(RegressionReport report)
+    {
+        if (report.CrashTestResult.CrashReport != null)
+        {
+            return await _taskOperations.GetByTaskId(report.CrashTestResult.CrashReport.TaskId);
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs:129]: In the python code we also use job id as part of the filter. Was that a query optimization?

**Review by stishkin — APPROVED** [review]

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs [hunk 6]
```diff
@@ -397,32 +394,74 @@ public record Scaleset(
...
     BlobRef? InputBlob,
-    string? Executable,
+    string Executable,
     string CrashType,
     string CrashSite,
-    List<string> CallStack,
-    string CallStackSha256,
-    string InputSha256,
+    IEnumerable<string> CallStack,
+    SHA256 CallStackSha256,
```
[the remark this answers is not in the document]
Comment (chkeita) [reply src/ApiService/ApiService/OneFuzzTypes/Model.cs:422]: SHA256 is class for computing hash. Can we use it to deserialize strings ?

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs [hunk 7]
```diff
@@ -397,32 +394,74 @@ public record Scaleset(
...
+    string? InputUrl,
     BlobRef? InputBlob,
-    string? Executable,
+    string Executable,
     string CrashType,
     string CrashSite,
-    List<string> CallStack,
-    string CallStackSha256,
-    string InputSha256,
+    IEnumerable<string> CallStack,
```
Comment (chkeita) [inline src/ApiService/ApiService/OneFuzzTypes/Model.cs:421]: Why the change to IEnumerable? All of these fields are of known size hen they are created. so LIstor even array would make more sense. We can generalize in the code instead of the data structure.

File: src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs (hunk 5 above)
Comment (chkeita) [reply src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs:121]: yes it is faster to query with both PartitionKey and RowKey than just RowKey

File: src/ApiService/ApiService/OneFuzzTypes/Events.cs [hunk 8]
```diff
@@ -249,25 +248,25 @@ PoolName PoolName
 //        NodeState state
 //        ) : BaseEvent();
 
-//    record EventCrashReported(
-//        Report Report,
-//        Container Container,
-//        [property: JsonPropertyName("filename")] String FileName,
-//        TaskConfig? TaskConfig
-//        ) : BaseEvent();
+record EventCrashReported(
```
Comment (chkeita) [inline src/ApiService/ApiService/OneFuzzTypes/Events.cs:251]: you will need to update GetventType and GetTypeInfo in BaseEvent to include these

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tevoinea) [reply src/ApiService/ApiService/OneFuzzTypes/Model.cs:422]: SHA256 is the wrong type, it's used for calculating hashes, not storing them

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs (hunk 7 above)
Comment (tevoinea) [reply src/ApiService/ApiService/OneFuzzTypes/Model.cs:421]: Good point, I switched it back

File: src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs (hunk 5 above)
Comment (tevoinea) [reply src/ApiService/ApiService/onefuzzlib/NotificationOperations.cs]: Changed it to use both job id and task id
